src/main/java/com/mraof/minestuck/world/gen/structure/wfc/WFC.java

```diff
@@ -0,0 +1,320 @@
+		
+		public void collapse(RandomSource random, PiecePlacer piecePlacer)
+		{
+			collapse(random, piecePlacer, null);
+		}
+		
+		public void collapse(RandomSource random, PiecePlacer piecePlacer, @Nullable WFCUtil.PerformanceMeasurer performanceMeasurer)
+		{
+			Set<WFCUtil.CellPos> cellsToGenerate = new HashSet<>(this.grid.availableEntriesMap.keySet());
+			
+			while(!cellsToGenerate.isEmpty())
+			{
+				if(performanceMeasurer != null)
+					performanceMeasurer.start(WFCUtil.PerformanceMeasurer.Type.ENTROPY_SEARCH);
+				WFCUtil.MinValueSearchResult<WFCUtil.CellPos> leastEntropyResult = WFCUtil.MinValueSearchResult.search(cellsToGenerate,
+						pos -> {
+							if(performanceMeasurer != null)
+								performanceMeasurer.start(WFCUtil.PerformanceMeasurer.Type.ENTROPY_CALC);
+							var entropy = this.grid.availableEntriesMap.get(pos).getEntropy();
+							if(performanceMeasurer != null)
+								performanceMeasurer.end(WFCUtil.PerformanceMeasurer.Type.ENTROPY_CALC);
+							return entropy;
+						});
+				if(performanceMeasurer != null)
+					performanceMeasurer.end(WFCUtil.PerformanceMeasurer.Type.ENTROPY_SEARCH);
+				
+				if(leastEntropyResult.entries().isEmpty())
+					break;
+				
+				WFCUtil.CellPos pos = Util.getRandom(leastEntropyResult.entries(), random);
+				cellsToGenerate.remove(pos);
+				
+				WFCUtil.EntropyList<WeightedEntry.Wrapper<WFCData.PieceEntry>> availableEntries = this.grid.availableEntriesMap.get(pos);
+				var chosenEntry = WeightedRandom.getRandomItem(random, availableEntries.entries());
+				if(chosenEntry.isEmpty())
+				{
+					piecePlacer.logNoEntries(pos);
+					continue;
+				}
+				
+				piecePlacer.place(pos, chosenEntry.get().data());
+				
+				if(performanceMeasurer != null)
+					performanceMeasurer.start(WFCUtil.PerformanceMeasurer.Type.ADJACENCY_UPDATE);
+				if(availableEntries.removeIf(entry -> entry != chosenEntry.get()))
+					this.grid.removeAdjacentUnsupportedEntries(pos, null);
+				if(performanceMeasurer != null)
+					performanceMeasurer.end(WFCUtil.PerformanceMeasurer.Type.ADJACENCY_UPDATE);
+			}
+		}
+	}
+	
+	public interface PiecePlacer
+	{
+		PiecePlacer EMPTY = new PiecePlacer()
+		{
+			@Override
+			public void place(WFCUtil.CellPos cellPos, WFCData.PieceEntry entry)
+			{
+			}
+			
+			@Override
+			public void logNoEntries(WFCUtil.CellPos cellPos)
+			{
+			}
+		};
+		
+		static PiecePlacer placeAt(WFCUtil.PositionTransform transform, StructurePieceAccessor pieceAccessor)
+		{
+			return new PiecePlacer()
+			{
+				@Override
+				public void place(WFCUtil.CellPos cellPos, WFCData.PieceEntry entry)
+				{
+					BlockPos pos = transform.toBlockPos(cellPos);
+					StructurePiece piece = entry.constructor().apply(pos);
+					if(piece != null)
+						pieceAccessor.addPiece(piece);
+				}
+				
+				@Override
+				public void logNoEntries(WFCUtil.CellPos cellPos)
+				{
+					BlockPos pos = transform.toBlockPos(cellPos);
+					WFC.LOGGER.warn("No entries possible at {}!", pos);
```

Comment:
Will recommend that this be turned into a debug message or hiding behind a config. Can cause a lot of clutter

Reply:
Principally, I would disagree with changing the log level as this is a very concrete hint towards a problem with the palette.

Practically, I don't think it matters right now. If the WFC is integrated and used properly later and there's a lot of clutter, we should primarily fix the WFC palette. If the problems at that point are acceptable but there is still clutter, then I'd be more fine with changing this at that point.